docs/RFCs/19042021_Cert_Management_In_CRDB_Helm_Chart.md

```diff
@@ -0,0 +1,243 @@
+* Contents of the Node certificate will be stored in the Node secret `cockroachdb-node`.
+* An annotation `managed-by: cockroachdb` will be added on secret
+* Node secret will be patched with annotation `resourceVersion` with the value of its current resourceVersion
+* Node secret will be patched with annotation `creationTime` and `duration` with current UTC time and value from `tls.certs.generate.nodeCertDuration`
+ 
+#### Generate-Client-Cert
+* A Client certificate will be generated by signing it with the generated CA.
+* Expiry of the certificate will be driven by the Helm value for `tls.certs.generate.clientCertDuration`, passed as env variable
+* Contents of the Client certificate will be stored in the Client secret `cockroachdb-root`
+* An annotation `managed-by: cockroachdb` will be added on secret
+* Client secret will be patched with annotation `resourceVersion` with the value of its current resourceVersion
+* Client secret will be patched with annotation `creationTime` and `duration` with current UTC time and value from `tls.certs.generate.clientCertDuration`
+ 
+#### Annotate-CA
+* CA secret will be patched with annotation `resourceVersion` with the value of its current resourceVersion
+* CA secret will be patched with annotation `creationTime` and `duration` with current UTC time and value from `tls.certs.generate.caCertDuration`
+* Now that CA is annotated with required info, it will be followed by [Node cert generation workflow](#generate-node-cert)
+ 
+#### Rotate-CA
+* CA rotation will follow the same workflow as [CA Generation Workflow](#generate-ca), except the new CA is a combined CA certificate that contains 
+  the new CA certificate followed by the old CA certificate.
+* This new combined CA certificate will be updated in the CA secret and all the Node and Client secret.
+  `TODO: Need to identify how to generate the SIGHUP signal in all the Nodes for certificate renewal`
+* In addition, Client secret and Node secret will be patched with annotation `needsRegeneration: True`, which specifies that their certificates
+  need to be regenerated in the next cronjob run. This is done in accordance with the suggestion in CockroachDB [doc](https://www.cockroachlabs.com/docs/v20.2/rotate-certificates.html#why-rotate-ca-certificates-in-advance)
+ 
+#### Check-node-cert-for-regeneration
+* Follow checks from [Check cert for regeneration](#check-cert-for-regeneration)
+* In addition, check if annotation `needsRegeneration: True` exists, if yes return True (This will be case when CA certificate has rotated, but Client and node certs are still to be recreated)
+ 
+#### Check-cert-for-regeneration
+* Check if the certificate secret has annotation `managed-by: cockroachdb`, if not return False
+* Check if the certificate is expiring in the next 2 months using the values from annotations `creationTime` and `duration` on secret, if expiring return True, else False
```

Comment:
Lets make the window where expiring is true configurable as well. `expiryWindow` or something

Reply:
The expiryWindow configurable is not straight forward becuase it leads to many edge cases of certificate rotation. For example:

Lets take three types of certificate and their duration:

| Client          |            Node            |              CA     |
|-----------------|---------------------------|--------------------|
|1 Year          |           2 Years         |           3 Year   |

Expiry window is 2 months (lets say provided by user)

1. The client certificate will rotate with first cron on 10th Month and again with second cron on 20th Month. 
2. During the second cron client certificate will be renewed but as the node certificate is going to expire before next cron, it will try to rotate node certificate on the 20th month. But if this node certificate is rotated with 2 year again it will outlive CA by 1 year, which is not valid.  

We have to device a method/formula with certain constraints to properly handle cert rotation.

One formula/approach could be:
We can have two crons having different schedules, one handle the CA rotation and other handles the client/node certificate rotation.

1. The CA cron schedule should be :  `Ca CRON schedule = (CA duration - CA expiryWindow)`
2. The cert rotation cron schedule should be:  `Cert rotation CRON =min((Node duration-Node expiryWindow), (client duration-client expiryWindow))`
3. The CA expiry window should be greater than the Cert rotation cron schedule:  `CA expiryWindow > Cert rotation CRON schedule`

Reply:
My only comment here is that we need a way to handle arbitrary user/client certs - we always generate a root cert, but if we are using tls auth each user needs their own cert.  Ideally, we could run a job to trigger the generation of an arbitrary user cert (e.g. `karl`) and have it also covered by this cron job.

Reply:
Updated the RFC doc with the latest changes as per our discussions. You can see the updated changes in this new [commit](https://github.com/cockroachdb/helm-charts/pull/80/commits/1f8b4555bc1843dbce07d7f3a8afd186b3c13336)